brainscore_language/models/huggingface.py

```diff
@@ -130,16 +130,16 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
     def estimate_reading_times(self, base_output):
         """
         :param base_output: the neural network's output
-        :return: perplexity (calculated as cross entropy) as a proxy for reading times
+        :return: cross entropy as a proxy for reading times, following Smith & Levy 2013
+            (https://www.sciencedirect.com/science/article/pii/S0010027713000413)
         """
         import torch.nn.functional as F
         tokens = self.tokenized_inputs['input_ids'].squeeze()
         logits = base_output.logits.squeeze()
```

Comment:
Right, this was missed in https://github.com/brain-score/language/pull/35.

How do you compute the reading time of a single word in that case? I.e. I show the subject the word "Hello", it takes them 1 second to read, how do you compare to the model? Language generation from scratch?

Reply:
Hi, I'm not sure I understand the question. Here's a slice of a reading time dataset:

![image](https://user-images.githubusercontent.com/6968/186253771-c5ec8fda-48ec-4096-97b4-f9a00003dcc4.png)

We compute a column `surprisal` by feeding the same text to a model that was used for an RT corpus. We index into the model's softmax outputs to get its estimate of $p(w_i | w_{< i})$ for each ground-truth word $w_i$. (There's some messiness re: tokens vs. words here of course, but that's the basic setup.)

You can see an example of the same indexing job in the [`SyntaxGym implementation`](https://huggingface.co/spaces/cpllab/syntaxgym/blob/main/syntaxgym.py#L197):

```python
        # Compute surprisal per-batch and combine into a single surprisal tensor.
        n_sentences, n_timesteps = all_tokenized_sentences["input_ids"].shape
        surprisals = torch.zeros(n_sentences, n_timesteps - 1).float().to(device)
        for i, batch in enumerate(datasets) :
            batch = batch.to(device)
            with torch.no_grad():
                # logits are B * T * V
                b_logits = model(batch)["logits"]
                b_surprisals = -b_logits.log_softmax(dim=2) / np.log(2)

            # Get surprisals of ground-truth words.
            gt_idxs = batch[:, 1:]
            # Reindexed surprisals: B * (T - 1)
            b_surprisals_gt = torch.gather(b_surprisals[:, :-1, :], 2, gt_idxs.unsqueeze(2)).squeeze(2)

            surprisals[i * batch_size : (i + 1) * batch_size] = b_surprisals_gt
```

Note the reindexing on the time axis of `gt_idxs = batch[:, 1:]` and `b_surprisals[:, :-1, :]`.

Reply:
ah so the first couple words in the sequence are actually ignored here? ("**If you were to** journey to the North of ...")

Reply:
That's not due to anything on the language modeling side -- that's because our RT regression model uses data about the surprisals of the past 4 words, which is of course undefined for the first 3 words in each corpus.

But it is true that we have to ditch the first token in the model's tokenized input, since by construction the model doesn't have a prediction for this item. But because models often have a beginning-of-sentence token, sometimes this doesn't matter.

Reply:
I'm a bit concerned about the model being able to output `nan` and then ignoring those predictions -- In an extreme case, the model could output all `nan`s except for a few words where it is very certain of the reading times and obtain a high score. My preference would be to output a non-nan value; we could do -1? That would signal non-regular output, but still allow it to be scored with a correlation metric without discarding values (that the model decides on).

Reply:
Copy-paste from Slack convo for posterity

> I see! That solves a problem but creates more, I think. The evaluation will be biased by any numeric value if we are using pearsonr or linear regression, such that the model results aren't interpretable (e.g. in a regression, the coefficient for surprisal outputs no longer really means surprisal).
As a benchmark dev I'd just pick through the model outputs and remove the -1 values manually so that my results are interpretable. Only difference then is that unaware benchmarkers might let this slip through without realizing.

> In any case models and benchmarks both will be doing data exclusions, right? Idk about the brain mapping, but in an RT benchmark for example, we usually drop 1) words at start and end of visual line, 2) words with punctuation. And models might reasonably want to drop OOV words (if that's a thing for them).
Is there a way we can track exclusions from each side as part of the benchmark record?

> In any eval I do I'm usually dropping thousands of tokens. Actually, in our latest RT paper (Wilcox 2020 CogSci) we dropped any token for which any model in our fleet did not provide valid surprisal estimates. That's probably the best option, but very hard to enforce on an evolving platform

> Suggested options summary
> 1. Explicitly track exclusions made by the model and by the benchmark; report this (e.g. as "coverage") alongside performance
> 2. Enforce that all models are evaluated on the same subset